Race in TestStorage/TestSingle

Unclear if the test or impl is buggy.  Details below.

```
2021-07-15T01:58:25.6111852Z ==================
2021-07-15T01:58:25.6112302Z WARNING: DATA RACE
2021-07-15T01:58:25.6112800Z Read at 0x00c000341d60 by goroutine 177:
2021-07-15T01:58:25.6113641Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*leasedReceiver).Close()
2021-07-15T01:58:25.6114687Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/leasedReceiver.go:85 +0xfb
2021-07-15T01:58:25.6115826Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*scheduler).Stop()
2021-07-15T01:58:25.6116826Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/scheduler.go:190 +0x253
2021-07-15T01:58:25.6118470Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*EventProcessorHost).Close()
2021-07-15T01:58:25.6119829Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/eph.go:375 +0x1ff
2021-07-15T01:58:25.6122007Z   github.com/Azure/azure-event-hubs-go/v3/storage.(*testSuite).TestSingle.func1()
2021-07-15T01:58:25.6123242Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/storage/eph_test.go:58 +0x95
2021-07-15T01:58:25.6124227Z   github.com/Azure/azure-event-hubs-go/v3/storage.(*testSuite).TestSingle()
2021-07-15T01:58:25.6125774Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/storage/eph_test.go:80 +0x719
2021-07-15T01:58:25.6126820Z   runtime.call32()
2021-07-15T01:58:25.6127597Z       /opt/hostedtoolcache/go/1.14.0/x64/src/runtime/asm_amd64.s:539 +0x3a
2021-07-15T01:58:25.6128395Z   reflect.Value.Call()
2021-07-15T01:58:25.6129116Z       /opt/hostedtoolcache/go/1.14.0/x64/src/reflect/value.go:321 +0xd3
2021-07-15T01:58:25.6129771Z   github.com/stretchr/testify/suite.Run.func1()
2021-07-15T01:58:25.6130466Z       /home/vsts/work/1/s/work/pkg/mod/github.com/stretchr/testify@v1.6.1/suite/suite.go:158 +0x391
2021-07-15T01:58:25.6131129Z   testing.tRunner()
2021-07-15T01:58:25.6132066Z       /opt/hostedtoolcache/go/1.14.0/x64/src/testing/testing.go:992 +0x1eb
2021-07-15T01:58:25.6132436Z 
2021-07-15T01:58:25.6132924Z Previous write at 0x00c000341d60 by goroutine 151:
2021-07-15T01:58:25.6133810Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*leasedReceiver).Run.func1()
2021-07-15T01:58:25.6135207Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/leasedReceiver.go:63 +0x91
2021-07-15T01:58:25.6135717Z 
2021-07-15T01:58:25.6136419Z Goroutine 177 (running) created at:
2021-07-15T01:58:25.6137322Z   testing.(*T).Run()
2021-07-15T01:58:25.6140091Z       /opt/hostedtoolcache/go/1.14.0/x64/src/testing/testing.go:1043 +0x660
2021-07-15T01:58:25.6141029Z   github.com/stretchr/testify/suite.runTests()
2021-07-15T01:58:25.6142061Z       /home/vsts/work/1/s/work/pkg/mod/github.com/stretchr/testify@v1.6.1/suite/suite.go:203 +0xe6
2021-07-15T01:58:25.6184362Z   github.com/stretchr/testify/suite.Run()
2021-07-15T01:58:25.6185922Z       /home/vsts/work/1/s/work/pkg/mod/github.com/stretchr/testify@v1.6.1/suite/suite.go:176 +0x8c3
2021-07-15T01:58:25.6187173Z   github.com/Azure/azure-event-hubs-go/v3/storage.TestStorage()
2021-07-15T01:58:25.6188189Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/storage/credential_test.go:48 +0x5e
2021-07-15T01:58:25.6188801Z   testing.tRunner()
2021-07-15T01:58:25.6189346Z       /opt/hostedtoolcache/go/1.14.0/x64/src/testing/testing.go:992 +0x1eb
2021-07-15T01:58:25.6189652Z 
2021-07-15T01:58:25.6190051Z Goroutine 151 (running) created at:
2021-07-15T01:58:25.6190814Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*leasedReceiver).Run()
2021-07-15T01:58:25.6191904Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/leasedReceiver.go:61 +0x1d5
2021-07-15T01:58:25.6192771Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*scheduler).startReceiver()
2021-07-15T01:58:25.6193650Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/scheduler.go:233 +0x486
2021-07-15T01:58:25.6194488Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*scheduler).scan()
2021-07-15T01:58:25.6195361Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/scheduler.go:126 +0x10bb
2021-07-15T01:58:25.6196389Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*scheduler).Run()
2021-07-15T01:58:25.6197287Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/scheduler.go:87 +0x181
2021-07-15T01:58:25.6198224Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*EventProcessorHost).StartNonBlocking.func1()
2021-07-15T01:58:25.6199439Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/eph.go:348 +0xf0
2021-07-15T01:58:25.6200146Z ==================
2021-07-15T01:58:25.6200620Z ==================
2021-07-15T01:58:25.6201230Z WARNING: DATA RACE
2021-07-15T01:58:25.6201731Z Write at 0x00c000071790 by goroutine 177:
2021-07-15T01:58:25.6202259Z   sync/atomic.CompareAndSwapInt32()
2021-07-15T01:58:25.6202838Z       /opt/hostedtoolcache/go/1.14.0/x64/src/runtime/race_amd64.s:293 +0xb
2021-07-15T01:58:25.6203412Z   sync.(*Mutex).Lock()
2021-07-15T01:58:25.6203946Z       /opt/hostedtoolcache/go/1.14.0/x64/src/sync/mutex.go:74 +0x49
2021-07-15T01:58:25.6204500Z   context.(*cancelCtx).cancel()
2021-07-15T01:58:25.6205235Z       /opt/hostedtoolcache/go/1.14.0/x64/src/context/context.go:395 +0x6c
2021-07-15T01:58:25.6205953Z   context.WithCancel.func1()
2021-07-15T01:58:25.6206871Z       /opt/hostedtoolcache/go/1.14.0/x64/src/context/context.go:235 +0x78
2021-07-15T01:58:25.6208954Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*leasedReceiver).Close()
2021-07-15T01:58:25.6210196Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/leasedReceiver.go:86 +0x1f8
2021-07-15T01:58:25.6211505Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*scheduler).Stop()
2021-07-15T01:58:25.6212462Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/scheduler.go:190 +0x253
2021-07-15T01:58:25.6213407Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*EventProcessorHost).Close()
2021-07-15T01:58:25.6214375Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/eph.go:375 +0x1ff
2021-07-15T01:58:25.6215314Z   github.com/Azure/azure-event-hubs-go/v3/storage.(*testSuite).TestSingle.func1()
2021-07-15T01:58:25.6216299Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/storage/eph_test.go:58 +0x95
2021-07-15T01:58:25.6217245Z   github.com/Azure/azure-event-hubs-go/v3/storage.(*testSuite).TestSingle()
2021-07-15T01:58:25.6219487Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/storage/eph_test.go:80 +0x719
2021-07-15T01:58:25.6220202Z   runtime.call32()
2021-07-15T01:58:25.6220786Z       /opt/hostedtoolcache/go/1.14.0/x64/src/runtime/asm_amd64.s:539 +0x3a
2021-07-15T01:58:25.6221518Z   reflect.Value.Call()
2021-07-15T01:58:25.6222064Z       /opt/hostedtoolcache/go/1.14.0/x64/src/reflect/value.go:321 +0xd3
2021-07-15T01:58:25.6222670Z   github.com/stretchr/testify/suite.Run.func1()
2021-07-15T01:58:25.6223345Z       /home/vsts/work/1/s/work/pkg/mod/github.com/stretchr/testify@v1.6.1/suite/suite.go:158 +0x391
2021-07-15T01:58:25.6223936Z   testing.tRunner()
2021-07-15T01:58:25.6224421Z       /opt/hostedtoolcache/go/1.14.0/x64/src/testing/testing.go:992 +0x1eb
2021-07-15T01:58:25.6224732Z 
2021-07-15T01:58:25.6225327Z Previous write at 0x00c000071790 by goroutine 151:
2021-07-15T01:58:25.6225802Z   context.WithCancel()
2021-07-15T01:58:25.6226297Z       /opt/hostedtoolcache/go/1.14.0/x64/src/context/context.go:233 +0xa7
2021-07-15T01:58:25.6227164Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*leasedReceiver).Run.func1()
2021-07-15T01:58:25.6228104Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/leasedReceiver.go:62 +0x5a
2021-07-15T01:58:25.6228580Z 
2021-07-15T01:58:25.6229044Z Goroutine 177 (running) created at:
2021-07-15T01:58:25.6229486Z   testing.(*T).Run()
2021-07-15T01:58:25.6229998Z       /opt/hostedtoolcache/go/1.14.0/x64/src/testing/testing.go:1043 +0x660
2021-07-15T01:58:25.6230543Z   github.com/stretchr/testify/suite.runTests()
2021-07-15T01:58:25.6231160Z       /home/vsts/work/1/s/work/pkg/mod/github.com/stretchr/testify@v1.6.1/suite/suite.go:203 +0xe6
2021-07-15T01:58:25.6231862Z   github.com/stretchr/testify/suite.Run()
2021-07-15T01:58:25.6232400Z       /home/vsts/work/1/s/work/pkg/mod/github.com/stretchr/testify@v1.6.1/suite/suite.go:176 +0x8c3
2021-07-15T01:58:25.6233302Z   github.com/Azure/azure-event-hubs-go/v3/storage.TestStorage()
2021-07-15T01:58:25.6234441Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/storage/credential_test.go:48 +0x5e
2021-07-15T01:58:25.6235082Z   testing.tRunner()
2021-07-15T01:58:25.6235618Z       /opt/hostedtoolcache/go/1.14.0/x64/src/testing/testing.go:992 +0x1eb
2021-07-15T01:58:25.6235978Z 
2021-07-15T01:58:25.6236414Z Goroutine 151 (running) created at:
2021-07-15T01:58:25.6237191Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*leasedReceiver).Run()
2021-07-15T01:58:25.6238587Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/leasedReceiver.go:61 +0x1d5
2021-07-15T01:58:25.6239487Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*scheduler).startReceiver()
2021-07-15T01:58:25.6240421Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/scheduler.go:233 +0x486
2021-07-15T01:58:25.6241577Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*scheduler).scan()
2021-07-15T01:58:25.6242562Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/scheduler.go:126 +0x10bb
2021-07-15T01:58:25.6243514Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*scheduler).Run()
2021-07-15T01:58:25.6244479Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/scheduler.go:87 +0x181
2021-07-15T01:58:25.6245502Z   github.com/Azure/azure-event-hubs-go/v3/eph.(*EventProcessorHost).StartNonBlocking.func1()
2021-07-15T01:58:25.6246404Z       /home/vsts/work/1/s/work/src/github.com/Azure/azure-event-hubs-go/eph/eph.go:348 +0xf0
2021-07-15T01:58:25.6246941Z ==================
2021-07-15T01:58:25.6247442Z     TestStorage/TestSingle: testing.go:906: race detected during execution of test
2021-07-15T01:58:25.6248588Z     TestStorage: suite.go:224: "goehtest7w57sm-tagpq4sp" does not contain "tagvpsn9", so it won't be deleted.
2021-07-15T01:58:25.6249396Z     TestStorage: testing.go:906: race detected during execution of test
2021-07-15T01:58:25.6250210Z --- FAIL: TestStorage (78.75s)
```

--- a/changelog.md
+++ b/changelog.md
@@ -1,5 +1,9 @@
 # Change Log
 
+## `v3.3.17`
+
+- Fixing issue where the LeaserCheckpointer could fail with a "ContainerAlreadyExists" error. (#253)
+
 ## `v3.3.16`
 
 - Exporting a subset of AMQP message properties for the Dapr project.

--- a/eng/azure-pipelines.yml
+++ b/eng/azure-pipelines.yml
@@ -41,7 +41,7 @@ jobs:
           go get github.com/AlekSi/gocov-xml
           go get -u github.com/matm/gocov-html
           go get -u golang.org/x/lint/golint
-          go get github.com/fzipp/gocyclo/cmd/gocyclo
+          go get github.com/fzipp/gocyclo/cmd/gocyclo@v0.3.1
         workingDirectory: '$(sdkPath)'
         displayName: 'Install Dependencies'
       - script: |

--- a/eng/integration-tests.yml
+++ b/eng/integration-tests.yml
@@ -29,7 +29,7 @@ steps:
       go get github.com/axw/gocov/gocov
       go get github.com/AlekSi/gocov-xml
       go get -u github.com/matm/gocov-html
-      go get github.com/fzipp/gocyclo/cmd/gocyclo
+      go get github.com/fzipp/gocyclo/cmd/gocyclo@v0.3.1
       go get golang.org/x/lint/golint
     displayName: 'Install Dependencies'
   - script: |

--- a/eng/integration-tests.yml
+++ b/eng/integration-tests.yml
@@ -47,10 +47,10 @@ steps:
       gocov-html < coverage.json > coverage.html
     displayName: 'Run Integration Tests'
     env:
-      ARM_SUBSCRIPTION_ID: $(go-live-azure-subscription-id)
-      ARM_CLIENT_ID: $(go-live-eh-azure-client-id)
-      ARM_CLIENT_SECRET: $(go-live-eh-azure-client-secret)
-      ARM_TENANT_ID: $(go-live-tenant-id)
+      ARM_SUBSCRIPTION_ID: $(aad-azure-sdk-test-subscription-id)
+      ARM_CLIENT_ID: $(aad-azure-sdk-test-client-id)
+      ARM_CLIENT_SECRET: $(aad-azure-sdk-test-client-secret)
+      ARM_TENANT_ID: $(aad-azure-sdk-test-tenant-id)
 
   - task: PublishTestResults@2
     inputs:

--- a/eng/integration-tests.yml
+++ b/eng/integration-tests.yml
@@ -47,7 +47,7 @@ steps:
       gocov-html < coverage.json > coverage.html
     displayName: 'Run Integration Tests'
     env:
-      ARM_SUBSCRIPTION_ID: $(aad-azure-sdk-test-subscription-id)
+      ARM_SUBSCRIPTION_ID: $(azure-subscription-id)
       ARM_CLIENT_ID: $(aad-azure-sdk-test-client-id)
       ARM_CLIENT_SECRET: $(aad-azure-sdk-test-client-secret)
       ARM_TENANT_ID: $(aad-azure-sdk-test-tenant-id)

--- a/eph/leasedReceiver.go
+++ b/eph/leasedReceiver.go
@@ -31,7 +31,7 @@ import (
 
 	"github.com/devigned/tab"
 
-	"github.com/Azure/azure-event-hubs-go/v3"
+	eventhub "github.com/Azure/azure-event-hubs-go/v3"
 )
 
 type (
@@ -58,11 +58,10 @@ func (lr *leasedReceiver) Run(ctx context.Context) error {
 	epoch := lr.lease.GetEpoch()
 	lr.dlog(ctx, "running...")
 
-	go func() {
-		ctx, done := context.WithCancel(context.Background())
-		lr.done = done
-		lr.periodicallyRenewLease(ctx)
-	}()
+	renewLeaseCtx, cancelRenewLease := context.WithCancel(context.Background())
+	lr.done = cancelRenewLease
+
+	go lr.periodicallyRenewLease(renewLeaseCtx)
 
 	opts := []eventhub.ReceiveOption{eventhub.ReceiveWithEpoch(epoch)}
 	if lr.processor.consumerGroup != "" {

--- a/storage/storage.go
+++ b/storage/storage.go
@@ -180,9 +180,8 @@ func (sl *LeaserCheckpointer) EnsureStore(ctx context.Context) error {
 			var storageErr azblob.StorageError
 
 			if errors.As(err, &storageErr) {
-				// if it already exists that's fine. There's a window where we can check that the container
-				// is there and then actually create it - any other consumer can beat us and it's fine.
-
+				// we're okay if the container has been created - we're basically racing against
+				// other LeaserCheckpointers.
 				if storageErr.ServiceCode() != azblob.ServiceCodeContainerAlreadyExists {
 					return err
 				}

--- a/storage/storage.go
+++ b/storage/storage.go
@@ -37,7 +37,7 @@ import (
 	"github.com/Azure/azure-amqp-common-go/v3/uuid"
 	"github.com/devigned/tab"
 
-	"github.com/Azure/azure-event-hubs-go/v3"
+	eventhub "github.com/Azure/azure-event-hubs-go/v3"
 	"github.com/Azure/azure-event-hubs-go/v3/eph"
 	"github.com/Azure/azure-event-hubs-go/v3/persist"
 
@@ -175,9 +175,22 @@ func (sl *LeaserCheckpointer) EnsureStore(ctx context.Context) error {
 	if !ok {
 		containerURL := sl.serviceURL.NewContainerURL(sl.containerName)
 		_, err := containerURL.Create(ctx, azblob.Metadata{}, azblob.PublicAccessNone)
+
 		if err != nil {
-			return err
+			var storageErr azblob.StorageError
+
+			if errors.As(err, &storageErr) {
+				// if it already exists that's fine. There's a window where we can check that the container
+				// is there and then actually create it - any other consumer can beat us and it's fine.
+
+				if storageErr.ServiceCode() != azblob.ServiceCodeContainerAlreadyExists {
+					return err
+				}
+			} else {
+				return err
+			}
 		}
+
 		sl.containerURL = &containerURL
 	}
 	return nil

--- a/storage/storage.go
+++ b/storage/storage.go
@@ -30,6 +30,7 @@ import (
 	"encoding/json"
 	"errors"
 	"io/ioutil"
+	"net/http"
 	"net/url"
 	"sync"
 	"time"
@@ -144,20 +145,22 @@ func (sl *LeaserCheckpointer) StoreExists(ctx context.Context) (bool, error) {
 	span, ctx := startConsumerSpanFromContext(ctx, "storage.LeaserCheckpointer.StoreExists")
 	defer span.End()
 
-	opts := azblob.ListContainersSegmentOptions{
-		Prefix: sl.containerName,
-	}
-	res, err := sl.serviceURL.ListContainersSegment(ctx, azblob.Marker{}, opts)
-	if err != nil {
-		return false, err
+	containerURL := sl.serviceURL.NewContainerURL(sl.containerName)
+	_, err := containerURL.GetProperties(ctx, azblob.LeaseAccessConditions{})
+
+	if err == nil {
+		return true, nil
 	}
 
-	for _, container := range res.ContainerItems {
-		if container.Name == sl.containerName {
-			return true, nil
+	var respErr azblob.ResponseError
+
+	if errors.As(err, &respErr) {
+		if respErr.Response().StatusCode == http.StatusNotFound {
+			return false, nil
 		}
 	}
-	return false, nil
+
+	return false, err
 }
 
 // EnsureStore creates the container if it does not exist

--- a/storage/storage_test.go
+++ b/storage/storage_test.go
@@ -25,6 +25,7 @@ package storage
 import (
 	"context"
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/Azure/azure-amqp-common-go/v3/aad"
@@ -32,6 +33,7 @@ import (
 	"github.com/stretchr/testify/assert"
 
 	"github.com/Azure/azure-event-hubs-go/v3/eph"
+	"github.com/Azure/azure-event-hubs-go/v3/internal/test"
 )
 
 const (
@@ -60,6 +62,35 @@ func (ts *testSuite) TestLeaserStoreCreation() {
 	ts.Require().NoError(err)
 
 	exists, err = leaser.StoreExists(ctx)
+	ts.NoError(err)
+	ts.True(exists)
+}
+
+func (ts *testSuite) TestLeaserStoreCreationConcurrent() {
+	wg := sync.WaitGroup{}
+
+	containerName := test.RandomString("concurrent-container", 4)
+
+	// do a simple test that ensures we don't die just because we raced with
+	// other leasers to create the storage container.
+	for i := 0; i < 100; i++ {
+		wg.Add(1)
+
+		go func(i int) {
+			defer wg.Done()
+
+			leaser, _ := ts.newLeaserWithContainerName(containerName)
+
+			err := leaser.EnsureStore(context.Background())
+			ts.Require().NoError(err)
+		}(i)
+	}
+
+	wg.Wait()
+
+	leaser, del := ts.newLeaserWithContainerName(containerName)
+	defer del()
+	exists, err := leaser.StoreExists(context.Background())
 	ts.NoError(err)
 	ts.True(exists)
 }
@@ -189,6 +220,10 @@ func (ts *testSuite) leaserWithEPH() (*LeaserCheckpointer, func()) {
 
 func (ts *testSuite) newLeaser() (*LeaserCheckpointer, func()) {
 	containerName := strings.ToLower(ts.RandomName("stortest", 4))
+	return ts.newLeaserWithContainerName(containerName)
+}
+
+func (ts *testSuite) newLeaserWithContainerName(containerName string) (*LeaserCheckpointer, func()) {
 	cred, err := NewAADSASCredential(ts.SubscriptionID, ts.ResourceGroupName, ts.AccountName, containerName, AADSASCredentialWithEnvironmentVars())
 	ts.Require().NoError(err)
 	leaser, err := NewStorageLeaserCheckpointer(cred, ts.AccountName, containerName, ts.Env)

--- a/version.go
+++ b/version.go
@@ -2,5 +2,5 @@ package eventhub
 
 const (
 	// Version is the semantic version number
-	Version = "3.3.13"
+	Version = "3.3.17"
 )

